designcompose/src/main/java/com/android/designcompose/DesignFrame.kt

```diff
@@ -560,17 +565,44 @@ internal fun Modifier.frameRender(
     document: DocContent,
     name: String,
     customizations: CustomizationContext,
+    maskInfo: MaskInfo?,
 ): Modifier =
     this.then(
         Modifier.drawWithContent {
-            render(
-                this@frameRender,
-                style,
-                frameShape,
-                customImageWithContext,
-                document,
-                name,
-                customizations,
-            )
+            fun render() =
+                render(
+                    this@frameRender,
+                    style,
+                    frameShape,
+                    customImageWithContext,
+                    document,
+                    name,
+                    customizations,
+                )
+            when (maskInfo?.type ?: MaskViewType.None) {
+                MaskViewType.MaskNode -> {
+                    // When rendering a mask, call saveLayer with blendmode DSTIN so that we blend
```

Comment:
oh, so we can't just make the mask the size of the mask node without having incorrect output, and there's no way to pass parentSize from parent to child because it isn't known until layout, which necessarily happens after the view tree composition...

I think it'd be possible to make a layout modifier that stuffs the parent size value into a more local mutableStateOf (so only the mask view needs the extra state & modifier, and the parent DesignView can be unaware), but I don't think it's worth the effort unless we find out that the arrangement here makes relayout slow.

Reply:
> oh, so we can't just make the mask the size of the mask node without having incorrect output, and there's no way to pass parentSize from parent to child because it isn't known until layout, which necessarily happens after the view tree composition...
> 
> I think it'd be possible to make a layout modifier that stuffs the parent size value into a more local mutableStateOf (so only the mask view needs the extra state & modifier, and the parent DesignView can be unaware), but I don't think it's worth the effort unless we find out that the arrangement here makes relayout slow.

Yea this was the only bit of complexity I ran into. That's an interesting idea though, will definitely look into that for the future.